clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -752,6 +758,106 @@ private StoragePartitionConfig getStoragePartitionConfig(
+        for (PubSubMessageHeader header: pubSubMessageHeaders.toList()) {
+          if (header.key().equals(VENICE_LEADER_COMPLETION_STATE_HEADER)) {
+            newState = LeaderCompleteState.valueOf(header.value()[0]);
+            partitionConsumptionState
+                .setLastLeaderCompleteStateUpdateInMs(consumerRecord.getValue().producerMetadata.messageTimestamp);
+            isLeaderCompleteHeaderFound = true;
+            break; // only interested in this header here
+          }
+        }
+        if (!isLeaderCompleteHeaderFound) {
+          // reset LeaderCompleteState: If a leader originally sent this header but later is rolled back to a version
+          // that doesn't support this header or this TP gets a new leader which doesn't support this header yet.
+          newState = LEADER_COMPLETE_STATE_UNKNOWN;
+        }
+
+        if (oldState != newState) {
+          LOGGER.info(
+              "LeaderCompleteState for store {} version {} partition {} changed from {} to {}",
+              storeName,
+              versionNumber,
+              partitionConsumptionState.getPartition(),
+              oldState,
+              newState);
+          partitionConsumptionState.setLeaderCompleteState(newState);
+        } else {
+          LOGGER.debug(
+              "LeaderCompleteState for store {} version {} partition {} received from leader: {} and is unchanged from the previous state",
+              storeName,
+              versionNumber,
+              partitionConsumptionState.getPartition(),
+              newState);
+        }
+        if (!partitionConsumptionState.isFirstHeartBeatSOSReceived()) {
+          partitionConsumptionState.setFirstHeartBeatSOSReceived(true);
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks whether the lag is acceptable for hybrid stores
+   *
+   * If the instance is a standby or DaVinciClient: Also check if
+   * 1. leaderCompleteStatus header has been received and
+   * 2. leader was completed and
+   * 3. the last update time was within 5 seconds
+   */
+  protected boolean isLagAcceptableForHybridStore(PartitionConsumptionState pcs, long offsetLag, long offsetThreshold) {
+    boolean isLagAcceptable = (offsetLag <= offsetThreshold);
+
+    if (isHybridMode() && (isDaVinciClient || pcs.getLeaderFollowerState().equals(STANDBY))) {
+      if (!isLagAcceptable) {
+        return false;
+      }
+
+      // non AA stores have issues reading HB SOS from the RT leading to the standby replicas waiting for
+      // leader completion state header indefinitely, so disabling it until that issue is resolved
+      if (!(getServerConfig().isLeaderCompleteStateCheckEnabled() && this.isActiveActiveReplicationEnabled())) {
+        return true;
+      }
+
+      // wait for the first HB to know if the leader supports sending LeaderCompleteState or not
+      if (!pcs.isFirstHeartBeatSOSReceived()) {
+        return false;
+      }
```

Comment:
I believe this is the problem for one colo (non-AA store). If the replica on the canary node is follower replica, and the leader is on very old codes which don't even produce HB. This follower replica will never be online?

Reply:
I thought about it for some time, the idea of writing correct codes still prevail. Maybe this is the case for our prod environment, but maybe not for external users of Venice (if any). If they use to run on older version (without HB code), and suddenly upgrade to this code version on one canary node; my understanding is that followers on that canary node will be stuck forever.

I could think of 3 options:
1. change the default value of config "isLeaderCompleteStateCheckEnabled" to false; update the javaDoc of that config, saying that users should not enable this until release xxx is fully rolled out. Then on our side, we can override this config in venice-backend immediately (so the default for us is true)
2. make a major version bump, explicitly stating that this version is not compatible with previous version.
3. Compute the elapsed time from the time when TOPIC_SWITCH CM is created to now; if the time is still within "getLeaderCompleteStateCheckValidIntervalMs" (reuse this config), return lag not acceptable; otherwise, return "boolean isLagAcceptable"

What do you think?

Reply:
Thanks for the suggestions, Min. As per our discussions: dropping 3rd option as it falls back to the old version (not checking HB) if the follower is running slow, which is what we are trying to fix in the first place.  Going with the first option.